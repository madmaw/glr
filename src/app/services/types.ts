import { type ExpressionService } from './expression';
import { type LoggingService } from './logging';

export type Services = {
  readonly loggingService: LoggingService,
  readonly expressionService: ExpressionService,
};

export type ServicesDescriptor = {
  readonly loggingService: 'local',
  readonly expressionService: 'local',
};
