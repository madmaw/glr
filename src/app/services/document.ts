import {
  type Document,
  type DocumentId,
} from 'app/domain/model';

export type DocumentService = {
  getDocument(id: DocumentId): Promise<Document>,
};
