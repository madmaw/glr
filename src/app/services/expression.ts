type Success<Type extends string, Value> = {
  type: Type,
  value: Value,
};

export type Result =
  | Success<'string', string>
  | Success<'number', number>
  | Success<'boolean', boolean>
  | {
    type: 'bad_expression',
    expression: string,
  }
  | {
    type: 'unexpected_result',
    resultType: string,
  };

export type ExpressionService = {
  evaluateExpression(expression: string): Promise<Result>,
};
