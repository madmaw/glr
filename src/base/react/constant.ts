import { useRef } from 'react';

// TODO although this is a ref, because it's hidden in useConstant here
// eslint will demand that the result be included in the list of deps
export function useConstant<T>(value: T): T {
  const ref = useRef(value);
  return ref.current;
}

type DeferredValue<T> = {
  type: 'unset',
} | {
  type: 'set',
  value: T,
};

// TODO although this is a ref, because it's hidden in useConstant here
// eslint will demand that the result be included in the list of deps
export function useDeferredConstant<T>(factory: () => T): T {
  const ref = useRef<DeferredValue<T>>({
    type: 'unset',
  });
  if (ref.current.type === 'unset') {
    ref.current = {
      type: 'set',
      value: factory(),
    };
  }
  return ref.current.value;
}
