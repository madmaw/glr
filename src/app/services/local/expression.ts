import { type Result } from 'app/services/expression';
import { type LoggingService } from 'app/services/logging';
import { delay } from 'base/delay';

export class LocalExpressionService {
  constructor(
    private readonly loggingService: LoggingService,
    private readonly delayMillis: number,
  ) {
  }

  async evaluateExpression(expression: string): Promise<Result> {
    await delay(this.delayMillis);
    if (!expression.trim().startsWith('return')) {
      expression = 'return ' + expression;
    }
    try {
      const f = new Function(expression);
      const result = await f();
      if (typeof result === 'string') {
        return {
          type: 'string',
          value: result,
        };
      }
      if (typeof result === 'boolean') {
        return {
          type: 'boolean',
          value: result,
        };
      }
      if (typeof result === 'number') {
        return {
          type: 'number',
          value: result,
        };
      }
      return {
        type: 'unexpected_result',
        resultType: typeof result,
      };
    } catch (error) {
      this.loggingService.warn('user expression failed', {
        error,
      });
      return {
        type: 'bad_expression',
        expression,
      };
    }
  }
}
