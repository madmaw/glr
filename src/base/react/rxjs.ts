import {
  useCallback,
  useEffect,
  useMemo,
} from 'react';
import {
  fromEvent,
  type Observer,
  type OperatorFunction,
  type PartialObserver,
  Subject,
} from 'rxjs';
import { useConstant } from './constant';

export type PropsWithEvents<E, P> = {
  events?: PartialObserver<E>,
} & P;

export function useEventsCallback<
  Name extends keyof GlobalEventHandlersEventMap,
  T,
  E extends GlobalEventHandlersEventMap[Name] = GlobalEventHandlersEventMap[Name],
>(
  operator: OperatorFunction<E, T>,
  events: PartialObserver<T> | undefined,
): (e: E) => void {
  const subject = useMemo(function () {
    return new Subject<E>();
  }, []);

  useEffect(function () {
    const subscription = subject.pipe(operator).subscribe(events);
    return subscription.unsubscribe.bind(subscription);
  }, [
    subject,
    operator,
    events,
  ]);

  return useCallback(function (e: E) {
    subject.next(e);
  }, [subject]);
}

export function useEvents<
  Name extends keyof GlobalEventHandlersEventMap,
  T,
  E extends GlobalEventHandlersEventMap[Name] = GlobalEventHandlersEventMap[Name],
>(
  name: Name,
  operator: OperatorFunction<E, T>,
  events: PartialObserver<T> | undefined,
  target: HTMLElement | null,
) {
  useEffect(function () {
    if (target != null) {
      const observable = fromEvent<E>(target, name);
      const subscription = observable.pipe(operator).subscribe(events);
      return subscription.unsubscribe.bind(subscription);
    }
  }, [
    name,
    operator,
    events,
    target,
  ]);
}

export function useObserverPipe<From, To>(
  to: PartialObserver<To>,
  operator: OperatorFunction<From, To>,
): Observer<From> {
  const subject = useConstant(new Subject<From>());
  useEffect(function () {
    const subscription = subject.pipe(operator).subscribe(to);
    return subscription.unsubscribe.bind(subscription);
  }, [
    operator,
    subject,
    to,
  ]);
  return subject;
}
