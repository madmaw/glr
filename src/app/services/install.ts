import { rollup } from 'base/record';
import { createDeferredServices } from './deferred';
import { install as installLocal } from './local/install';
import {
  type Services,
  type ServicesDescriptor,
} from './types';

export function install({ descriptors }: { descriptors: ServicesDescriptor }): Services {
  const immediateServices: Partial<Services> = {};
  const deferredServices = createDeferredServices(immediateServices);

  const localServices = installLocal({
    services: deferredServices,
    descriptors,
  });
  return rollup(
    immediateServices,
    localServices,
  );
}
