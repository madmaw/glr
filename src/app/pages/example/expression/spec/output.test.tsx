import { ExpressionOutput } from 'app/pages/example/expression/output';
import { render } from 'app/pages/example/testing/react';
import { type Result } from 'app/services/expression';

suite('ExpressionOutput', function () {
  describe.each<[string, Result]>([
    [
      'number',
      {
        type: 'number',
        value: 12,
      },
    ],
    [
      'string',
      {
        type: 'string',
        value: 'asdf',
      },
    ],
    [
      'boolean',
      {
        type: 'boolean',
        value: true,
      },
    ],
    [
      'unexpected result type',
      {
        type: 'unexpected_result',
        resultType: 'a cat',
      },
    ],
    [
      'bad expression',
      {
        type: 'bad_expression',
        expression: 'asdf',
      },
    ],
  ])('%s', function (_, result) {
    test('renders', function () {
      const { container } = render(<ExpressionOutput result={result} />);
      expect(container).toMatchSnapshot();
    });
  });
});
