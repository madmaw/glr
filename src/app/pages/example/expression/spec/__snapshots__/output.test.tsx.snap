// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ExpressionOutput > bad expression > renders 1`] = `
<div>
  <div
    class="css-995cds"
    direction="row"
  >
    <svg
      fill="none"
      height="32"
      viewBox="0 0 36.95041722813605 32"
      width="36.95041722813605"
    >
      <mask
        id="mask"
      >
        <g>
          <rect
            fill="white"
            height="32"
            width="36.95041722813605"
            x="0"
            y="0"
          />
          <text
            alignment-baseline="middle"
            fill="black"
            font-size="25.6"
            text-anchor="middle"
            x="18.475208614068023"
            y="21.333333333333332"
          >
            !
          </text>
        </g>
      </mask>
      <g
        id="tri"
        mask="url(#mask)"
      >
        <path
          d="
            M 17.260677434251694 8
            A 5.333333333333333 5.333333333333333 0 0 1 26.498281741285705 8
            L 39.140157022020404 28.333333333333336
            A 5.333333333333333 5.333333333333333 0 0 1 34.521354868503394 36.333333333333336
            L 9.237604307034008 36.333333333333336
            A 5.333333333333333 5.333333333333333 0 0 1 4.618802153517002 28.333333333333336
            z
        "
          fill="rgb(0% 0% 0%)"
          stroke="none"
          transform="translate(-3.904270973700675, -5.333333333333335)"
        />
      </g>
    </svg>
    <span
      class="css-1tlx7i-typographic"
      type="1"
    >
      Invalid Expression
    </span>
  </div>
</div>
`;

exports[`ExpressionOutput > boolean > renders 1`] = `
<div>
  <span
    class="css-1tlx7i-typographic"
  >
    Boolean: true
  </span>
</div>
`;

exports[`ExpressionOutput > number > renders 1`] = `
<div>
  <span
    class="css-1tlx7i-typographic"
  >
    Number: 12
  </span>
</div>
`;

exports[`ExpressionOutput > string > renders 1`] = `
<div>
  <span
    class="css-1tlx7i-typographic"
  >
    String: asdf
  </span>
</div>
`;

exports[`ExpressionOutput > unexpected result type > renders 1`] = `
<div>
  <div
    class="css-995cds"
    direction="row"
  >
    <svg
      fill="none"
      height="32"
      viewBox="0 0 36.95041722813605 32"
      width="36.95041722813605"
    >
      <mask
        id="mask"
      >
        <g>
          <rect
            fill="white"
            height="32"
            width="36.95041722813605"
            x="0"
            y="0"
          />
          <text
            alignment-baseline="middle"
            fill="black"
            font-size="25.6"
            text-anchor="middle"
            x="18.475208614068023"
            y="21.333333333333332"
          >
            !
          </text>
        </g>
      </mask>
      <g
        id="tri"
        mask="url(#mask)"
      >
        <path
          d="
            M 17.260677434251694 8
            A 5.333333333333333 5.333333333333333 0 0 1 26.498281741285705 8
            L 39.140157022020404 28.333333333333336
            A 5.333333333333333 5.333333333333333 0 0 1 34.521354868503394 36.333333333333336
            L 9.237604307034008 36.333333333333336
            A 5.333333333333333 5.333333333333333 0 0 1 4.618802153517002 28.333333333333336
            z
        "
          fill="rgb(0% 0% 0%)"
          stroke="none"
          transform="translate(-3.904270973700675, -5.333333333333335)"
        />
      </g>
    </svg>
    <span
      class="css-1tlx7i-typographic"
      type="1"
    >
      Expression must return a primitive
    </span>
  </div>
</div>
`;
